### DISCUSSION TITLE
Do you know why Next.js is awesome? - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by piers-sinclair** [SSWConsulting/SSW.Rules.Content#5174] [Chewing the Fat]
Contributors: @wicksipedia @pierssinclairssw

Let's talk about this rule:
https://www.ssw.com.au/rules/use-nextjs/

**Comment by GordonBeeming** [comment]
I learned what's behind the name nextjs that I've heard before, never really looked into it but nice to get an overview of its usage of it.

*personal preference*
Js frameworks change every other month... nextjs feels like the next hype wave... .net isn't slow or a resource hog. My personal preference is straight .net, potentially blazor for a more flashy front end using existing skills. Also, with the right caching and using your hosting platform correctly, you can get your content cached on the edge and purged when it changes too.

**Comment by bradystroud** [comment]
Next.Js is cool, but most of the time I would still prefer Blazor.
It depends on the project.

**Comment by 2fernandez** [comment]
![4815823](https://user-images.githubusercontent.com/27670973/224163789-ff442441-4cc8-4f72-b0b2-888be566d8c1.jpg)

**Comment by matt-goldman** [comment]
These CTF questions are often framed in the context of "is this the next big thing and should we use this now for everything?".

I think this is the wrong question.

Next.js looks cool, and while I don't fully comprehend the use case (have we come full circle with web dev and gone back to what is essentially, albeit more sophisticated, ajax?), it seems to me like a good solution for _websites_. For an enterprise application I don't think it's the right option for SSW, and would still say that .NET API + Angular or .NET + Blazor is a better option.

**Comment by AlexRothwell** [comment]
The rule is a great starting point for understanding what Next.js is and I really appreciate the effort taken there. Being able to read a reasonably short intro so I can finally follow what people are talking about is something I find valuable

As for whether we should be going all in for a particular technology, I think continuing to have such a breadth of knowledge at SSW is important. It lets us contribute to a wide range of projects, whether they're greenfield or established

Personally I'm excited to see where Blazor United goes

**Comment by ozdemir-mehmet** [comment]
Gordon has said what I was going to say. Frameworks come and go, use what works at the time.

**Comment by calumjs** [comment]
I think next.js is awesome - I have been using it for a couple of years and it has only been getting better.
I also love my next.js mug:
![image](https://user-images.githubusercontent.com/13717390/224171067-28815d50-d4be-4080-b116-e993da879963.png)

**Comment by bahjatalaadelssw** [comment]
More Google juice is good

**Comment by tkapa** [comment]
NextJs is great. I MUCH prefer to use it over Angular. I think that when we're weighing options for new sites / revamps, NextJs and Blazor should be at the front of our mind.

**Comment by JackDuan123** [comment]
Next.JS is ok
But its API feature is wired, I don't like it. It is hard to maintain, it takes years to build.
The server-side render is pretty hard to use.

**Comment by tombui99** [comment]
Next.JS is awesome! It's fast, it's static which means we can use SSW CodeAuditor to run scan on it!

**Comment by cvionnet** [comment]
I have not used Next.js, but based on the features and capabilities presented in the video, it does seem interesting.
I do agree with other comments recommending .Net + Blazor as an alternative, depending on the project.

**Comment by ravidhaduk** [comment]
Haven't touched it. Seems cool but would still prefer Angular + .NET. 🙌

**Comment by christoment** [comment]
I like NextJS since its SSR/SSG/ISR feature is one of the best I've seen in JS world (even better than Gatsby - I've replaced my Gatsby blog with NextJS, it's super easy!).
I don't really like API part since JS is a loosely typed language so it would be a challenge to impose the right data format compared to .NET.

.NET  is still my preference for creating API :)

At the moment, I still recommend Angular for clients though since it has opinionated framework - and it's very useful in our field where people can come and go on different Angular project and they can still find their way around the project.

**Comment by AntPolkanov** [comment]
I've used Next.js before and it's super cool. It simplifies heaps of thigs like SSR, code splitting, dynamic import etc. All those things were a bit difficult to implement yourself before Next.js came out. It's a really good choice for corporate websites, blogs and online shops.

**Comment by SamBWagner** [comment]
I think that Font-End frameworks should simply be chosen based on internal skillset (How many of our devs know it?) and current ubiquity (How many other people use it?). JavaScript is still going to be JavaScript and we don't usually choose it for back-end work. This is, I assume, for a myriad of reasons and having a new front-end isn't going to be the straw that broke the camel's back.

**Comment by piers-sinclair** [comment]
Next.js look awesome and has a lot of big ideas.

It's an interesting concept, but I'll need to use it in practice before I make any definitive call.

**Comment by Harry-Ross** [comment]
I thought that this was a great introduction to NextJS, a very powerful framework for building frontend applications. I have used NextJS in the past for a couple personal projects, and thought it was great for SSR. I have found it to be a really great experience and I much prefer the developer experience compared with vanilla React. I believe NextJS is the clear choice for server side rendered React.

However, I'm not really a fan of using it as a backend framework, as the developer experience is nowhere near as good as most other specialised backend solutions.

**Comment by NadeekaJK** [comment]
Good stuff!! Looks awesome. Dynamic routing is not readily available, needs Node.js to set it up. Hope this gets fixed soon!!

**Comment by vladislav-kir** [comment]
Next.JS sounds like an interesting framework. It was nice to learn a little more about it.

**Comment by leehawkinsssw** [comment]
While I'm not a developer (at least not since my days writing LISP code in the mid-late 1990s), it seems to me that Next.js is designed to solve some kinds of problems that are not covered well by existing frameworks. 

I'm not a trend follower and it seems too early to go "all in" on this latest framework. It's important to keep an eye on trends in the fast-paced tech industry but I find it pays to wait until the genuinely useful and adopted stuff becomes clear. I liken this to what Tara Brach says about mindfulness meditation: "you are developing the capacity to recognize when thoughts are happening without getting lost in the story line". Watch the trends and let them pass by, the important ones will rise up and then it's worth your time and energy to dive deeper.

**Comment by JackDevAU** [comment]
I've been using Next.js for a while now and really do love it. (Even trying out the experimental versions when I have spare time). However, I disagree with a blanket statement of this is the framework we should use and really push for. It really depends on the project, what their devs are familiar with, what devs we have available, what they are looking to build, etc. I would still always prefer a .NET Api especially for enterprise level solutions, cause at the end of the day... Node is slow. But for a full featured react framework, Next.js is one of the best! 

I see above comments about the Api being confusing and weird to work with.... and it is a little weird at the moment. With Next.js 13 they are changing how the Api routes work and its looking to be a much better approach!

**Comment by camillars** [comment]
clearly not a dev, but if Next.js is better for SEO, I am a fan already! 🤩

**Comment by amankumarrr** [comment]
I recently started working on it and initially, I thought it was just a React thing. However, when it builds to the client side, the rendering is incredibly fast. One thing to note though is that with MDX files, you cannot play around with the indentation as it's quite sensitive.

**Comment by Ben0189** [comment]
Next.js looks great, provide interesting feature like ISR. However, I prefer Angular than Next.Js for medium-large application due to the scalabilities and code reusability provided by angular

**Comment by drwharris** [comment]
Next what?

I think I need to do some reading!

**Comment by steven0x51** [comment]
Frameworks evolve frequently, I believe there's no best framework for all projects. It's more important to choose the right framework for the project rather than to push a framework for all projects.

**Comment by UlyssesMaclaren** [comment]
sounds cool :)
I wonder what the Blazor equivalent would be

**Comment by Levijj22** [comment]
![CoolB99GIF](https://user-images.githubusercontent.com/115961605/224185750-d4793971-46f0-4fd7-8ecf-b33fb32959be.gif)

**Comment by apforsyth** [comment]
Love to see other examples of website build with nextjs

**Comment by KahaMason** [comment]
Next.js is awesome. I've used it in the past and I find it easy to use. I would prefer is over angular is most cases. I've always seen Next.js as a purely front-end framework and I found it interesting learning of some of it's backend capabilities. I still prefer .NET for handling APIs as my personal preference.

**Comment by Geordie88** [comment]
It was cool to get an overview of what next.js is but I am not knowledgeable enough about the alternatives and the pros and cons of different solutions to really weigh in on the discussion.

**Comment by AshAnil** [comment]
Next.js is a good choice if you need a lightweight, fast, and easy-to-use framework for building static or dynamic websites.

**Comment by stephanfakossw** [comment]
![2023-03-10_11-07-40](https://user-images.githubusercontent.com/96279310/224189509-34d3ba30-a0e2-4f04-8476-732812c1e9c9.jpg)

**Comment by Hona** [comment]
Next.js looks cool but I haven't yet had a chance to play with it!

**Comment by chrisschultzssw** [comment]
Not a dev, but I get the feeling the best framework to use is... it depends.

It is cool that this is a constantly evolving tech, where new frameworks being built to overcome limitations of existing solutions.

**Comment by ozair-ashfaque** [comment]
Next.js seems pretty awesome and it would be good if I learn more about it. But my preference is .NET API + Angular

**Comment by jimmy1992abc** [comment]
Too hard for me to understand, lol.
But looks like a revolutionary framework.

**Comment by landonmaxwell** [comment]
This would be a good relevant topic to make some videos around.

**Comment by danielmackay** [comment]
Next.JS is awesome!

In the past we have had to choose between a static site (Razor Pages), a SPA (Blazor), or a Frankenstein of something in-between.  The biggest thing I like about next.js is that you have the freedom choose between dynamic or static on a page by page basic.

When you're looking at Next.JS, Vercel is a natural progress and provides several benefits for running a next.js site.  Perhaps we need a rule or CTF on that also?

I do wonder if the Blazor Unified project is looking to go down a similar route to next.js and provide flexibility of either static or dynamic on a per page basis.  Will be a very interesting space to watch...

**Reply by wicksipedia** [reply]
I love that its all integrated together in next - I can have my api sitting in the same project that may aggregate other apis behind the firewall. 

As much as I love angular - I need to have feet in 2 camps to make things happen. I'm still a Blazor skeptic 😬

**Reply by danielmackay** [reply]
@wicksipedia - if you needed to build a large API that included a lot of data access to power a Next.JS site, what would be your preference for structuring that?

Option 1 - Go all in on next.js and do everything in the same eco system?
Option 2 - Use next.js for all UI concerns and build a separate .NET API to do the heavy lifting?
Option 3 - Something else?

**Reply by wicksipedia** [reply]
I guess I'd still default to a .net api since I know it works and nswag can do a lot of boilerplate stuff for me.

I think the fact that I considered doing it in a javascript framework is a big deal by itself

**Comment by PennyWalker** [comment]
It's interesting to me to see what the devs think about new tech... I think there are always going to people who love it and people who hate it, that is certainly the case with Angular and React. Gordon probably said it best  - good to investigate new things, but choose the best one for your projects requirements.

**Comment by hveraus** [comment]
Do I have to learn ReactJs first before start with NextJS?

**Reply by danielmackay** [reply]
All the UI side is still based on ReactJS, so yes you do. 🙂

**Comment by jerryluossw** [comment]
Next.JS is cool, but not for all projects, e.g., Node.js is not suitable for big engineering projects.
I think there is no best technology, only the most suitable technology.

**Comment by AnkyChe** [comment]
Not a dev but it's great to learn. Seems Next.JS is an awesome framework.

**Comment by Berber-Mama** [comment]
It's interesting to know Next.Js!

**Comment by KristenHu** [comment]
NextJS is cool, there is no doubt. I am a beginner to this frame, but I still think it's pretty good. But for a different project, I think the solution also would be different, so NextJs is not always the best.

**Comment by sylhuang** [comment]
Interested in playing around with this framework. Haven't used it yet.

**Comment by SofieHong** [comment]
TOO hard for me to understand 🤐
But it looks like a great video topic that a lot of dev will be interested in 🤩

**Comment by johnxu16** [comment]
It is great to use NextJS as an out-of-box React website framework since NextJS supports a lot of different processes of building websites (CSR, SSR, SSG).
However, if we want to use it as a backend server as well, there are so many middlewares we have to apply and configure to make it robust and friendly to new developers.
There are some other better options like Ne**s**tJS.

**Comment by jimmidier** [comment]
It's been a pleasant journey to learn and use Next.JS in the new SSW Website project, with the only exception for the hydration error nightmare😢

**Comment by AttackOnMorty** [comment]
<img width="534" alt="image" src="https://user-images.githubusercontent.com/37203901/224211104-3bac8308-9c33-4844-b9be-47abf0268a6b.png">

From https://remix.run/blog/remix-vs-next

**Comment by leahy268** [comment]
Not really an area that I've looked into or know a lot about.

**Comment by zl24622** [comment]
Learnt Next.js can do server side rendering. 

Front end development is not my area, generally like keeping things separated and lightweight just for testability. So probably still go .Net for API. Unless the website is not heavily replying on data, then consider using the Next.js for both ends.

**Comment by PothieuG** [comment]
I have never used Next.js myself, but I have heard nothing but good things about it. 
Some of my past colleagues have tried it for their personal projects and never looked back! I plan to try it out myself in order to provide a proper review.

**Comment by Freego1783** [comment]
Clearly not informed enough as I've never used it, but it looks promising. I'll probably give it a try.

**Comment by BrookJeynes** [comment]
NextJS is great and has a lot of potential however, it shouldn't be used for everything. All tech has its pros and cons and needs to be analyzed for the project at hand.

**Comment by DavidAbraham082** [comment]
Looks pretty cool

**Comment by jernejk** [comment]
NextJS seems to be cool tech and I'm planning to use it for my blog instead of Ghost.
I haven't worked a lot on UI front other then Xamarin for a client and Blazor for my pet projects in the last 12 months, which is why I'm biased toward Blazor for internal enterprise apps and pet projects.
For public facing apps, JS heavy or lightweight app, I prefer JS frameworks and I'll look into NextJS for next project.

**Comment by TobyChurches** [comment]
NextJS seems like a good solution to use, however, without experience with it (or React) I still prefer the familiar .NET and Angular combo 😅

**Comment by billyterblanche** [comment]
It was cool to learn about Next.js a bit more as I've seen it pop up on my YouTube feed, but just never clicked on any of the videos.

TBH, I'm still a bit lost on SSR vs RSC and the impact it has on load time and SEO 😵
